art/attacks/poisoning/gradient_matching_attack.py

```diff
@@ -0,0 +1,315 @@
+                :num_poison_samples
+            ]
+            x_poison = x_train[indices_poison]
+            y_poison = y_train[indices_poison]
+            x_poisoned, B_ = poisoner(x_trigger, y_trigger, x_poison, y_poison)  # pylint: disable=C0103
+            B_ = np.mean(B_)  # Averaging B losses from multiple batches.  # pylint: disable=C0103
+            if B_ < best_B:
+                best_B = B_  # pylint: disable=C0103
+                best_x_poisoned = x_poisoned
+                best_indices_poison = indices_poison
+
+        x_train[best_indices_poison] = best_x_poisoned
+        return x_train, y_train  # y_train has not been modified.
+
+    def __poison__pytorch(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_poison: np.ndarray, y_poison: np.ndarray
+    ) -> np.ndarray:
+        raise NotImplementedError
+
+    def __poison__tensorflow(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_poison: np.ndarray, y_poison: np.ndarray
+    ) -> np.ndarray:
+        """
+        Optimize the poison by matching the gradient within the perturbation budget.
+
+        :param x_trigger: List of triggers.
+        :param y_trigger: List of target labels.
+        :param x_poison: List of samples to poison.
+        :param y_poison: List of the labels for x_poison.
+        :return: A pair of poisoned samples, B-score (cosine similarity of the gradients).
+        """
+        import tensorflow.keras.backend as K
+        import tensorflow as tf
+        from tensorflow.keras.layers import Input, Embedding, Add, Lambda
+
+        # Get the target gradient vector.
+        def grad_loss(model, x, target):
+            with tf.GradientTape() as t:  # pylint: disable=C0103
+                t.watch(model.weights)
+                output = model(x)
+                loss = model.compiled_loss(target, output)
+            d_w = t.gradient(loss, model.trainable_weights)
+            d_w = tf.concat([tf.reshape(d, [-1]) for d in d_w], 0)
+            d_w_norm = d_w / tf.sqrt(tf.reduce_sum(tf.square(d_w)))
+            return d_w_norm
+
+        grad_ws_norm = grad_loss(self.substitute_classifier.model, tf.constant(x_trigger), tf.constant(y_trigger))
+
+        class ClipConstraint(tf.keras.constraints.MaxNorm):
+            """
+            Clip the tensor values.
+            """
+
+            def __init__(self, max_value=2):
+                super().__init__(max_value=max_value)
+
+            def __call__(self, w):
+                return tf.clip_by_value(w, -self.max_value, self.max_value)
+
+        # Define the model to apply and optimize the poison.
+        input_poison = Input(batch_shape=self.substitute_classifier.model.input.shape)
+        input_indices = Input(shape=())
+        # y_true_poison = Input(shape=self.substitute_classifier.model.output.shape)
+        # y_true_poison = Input(shape=np.shape(y_trigger)[1:])
```

Comment:
Are these commented lines still needed?

Reply:
Not needed. I already removed those in my branch.. I will update the PR soon.